app/assets/javascripts/Components/submission_file_manager.jsx

```diff
@@ -114,22 +117,41 @@ class SubmissionFileManager extends React.Component {
         data.append("grouping_id", this.props.grouping_id);
       }
       data.append("unzip", unzip);
-      $.post({
+
+      $.ajax({
         url: Routes.update_files_course_assignment_submissions_path(
           this.props.course_id,
           this.props.assignment_id
         ),
+        type: "POST",
+        processData: false,
+        contentType: false,
         data: data,
-        processData: false, // tell jQuery not to process the data
-        contentType: false, // tell jQuery not to set contentType
+        xhr: () => {
+          const xhr = new XMLHttpRequest();
+
+          xhr.upload.addEventListener(
+            "progress",
+            event => {
+              if (event.lengthComputable) {
+                this.setState({uploadModalProgressPercentage: (event.loaded / event.total) * 100});
+              }
+            },
+            false
+          );
+
+          return xhr;
+        },
       })
         .then(typeof this.props.onChange === "function" ? this.props.onChange : this.fetchData)
         .then(this.endAction)
+        .then(() => this.setState({showUploadModal: false, uploadTarget: undefined}))
```

Comment:
This should be part of the state change under `always`; combine with the state setting for the progress bar. If there's an error message, e.g. if the file size is too large, there will be a flash message displayed and we want to make sure it's visible to users.